### ISSUE TITLE
Does Transformer use significantly less GPU memory than LSTM with comparable size?

### ISSUE THREAD

**Issue by AranKomat** [tensorflow/tensor2tensor#519]
Obviously, Transformer performs far better than LSTM in translation and some other tasks. I've also heard that it's much more memory-efficient: on a single 1080ti, one can train a batch size of 2048 with Transformer, while for a comparable LSTM, at most he can train a batch size of 64. I don't see why this is the case considering that the most memory hungry part is the embedding and softmax layers, which are the same for both architectures. I'd love to hear your thought and anecdote comparing memory usage of both architectures.

**Comment by martinpopel** [comment]
I am not sure how do you define "comparable" LSTM. Both self-attention (Transformer) and LSTM can have multiple layers and different hidden sizes etc.
Where do you got the number 64? I guess this is batch size in number of sentences (as used in most frameworks except for T2T), while batch_size=2048 in T2T is the number of subwords (which may corresponds e.g. to 64 sentences with average size 32 subwords).

**Comment by AranKomat** [comment]
This is a statement I found online [(here)](https://github.com/CR-Gjx/LeakGAN/issues/2), so I cannot be so sure. He was talking about the case of hidden_dim=128, so I assume he's talking about transformer_tiny or something with not so many layers (<8). Since LSTM doesn't usually have many layers, and since the claimed memory usage is different by 32 fold, what matters more is probably whether he meant the batch size as the total sentence number or the total token number as you meant. Anyhow, if you guys don't find a significant difference in GPU memory usage between LSTM and Transformer, that's the confirmation I wanted to get.
